File: feel-engine/src/main/java/org/camunda/feel/spi/JavaFunctionProvider.java
/* Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.camunda.feel.spi;

import java.util.Optional;

import org.camunda.feel.interpreter.Val;
import org.camunda.feel.interpreter.ValFunction;

import scala.Function1;
import scala.Option;
import scala.collection.JavaConverters;
import scala.collection.immutable.List;
import scala.runtime.AbstractFunction1;

public interface JavaFunctionProvider extends CustomFunctionProvider
{
    Optional<JavaFunction> resolveFunction(String functionName, int argCount);

    @Override
    default Option<ValFunction> getFunction(String functionName, int argCount)
    {
        final Optional<JavaFunction> function = resolveFunction(functionName, argCount);

        if (function == null || !function.isPresent())
        {
            return Option.empty();
        }
        else
        {
            return Option.apply(asFunction(function.get()));
        }
    }

    default ValFunction asFunction(final JavaFunction function)
    {
        final List<String> paramList = JavaConverters.asScalaBufferConverter(function.getParams()).asScala().toList();

        final Function1<List<Val>, Val> f = new AbstractFunction1<List<Val>, Val>()
        {
            public Val apply(List<Val> args) {

                final java.util.List<Val> argList = JavaConverters.<Val> bufferAsJavaListConverter(args.toBuffer()).asJava();

                return function.getFunction().apply(argList);
            };

        };

        return new ValFunction(paramList, f, function.isInputVariableRequired());
    }

}

File: feel-engine/src/main/scala/org/camunda/feel/spi/JavaFunctionProvider.scala
package org.camunda.feel.spi

import java.util.{Collections, Optional}
import org.camunda.feel.interpreter._
import scala.collection.JavaConverters._

/**
  * Provides one or more functions which can be used in an expression.
  */
abstract class JavaFunctionProvider extends CustomFunctionProvider {

  /**
    * Returns the function for the given name.
    *
    * @param functionName the name of the function
    * @return the function or [[Optional.empty()]], if no function is provided for this name
    */
  def resolveFunction(functionName: String): Optional[JavaFunction]

  /**
    * Returns a list of functions for the given name. There can be multiple functions with different parameters.
    *
    * @param functionName the name of the function
    * @return a list of functions or an empty list, if no function is provided for this name
    */
  def resolveFunctions(functionName: String): java.util.List[JavaFunction] = {
    val function = resolveFunction(functionName)

    if (function.isPresent) {
      Collections.singletonList(function.get)
    } else {
      Collections.emptyList()
    }
  }

  override def getFunctions(name: String): List[ValFunction] = {
    resolveFunctions(name).asScala.map(f => asFunction(f)).toList
  }

  override def getFunction(name: String): Option[ValFunction] = {
    getFunctions(name).headOption
  }

  private def asFunction(function: JavaFunction): ValFunction = {

    val paramList: List[String] = function.getParams().asScala.toList

    val f: (List[Val] => Val) = (args: List[Val]) => {

      val argList: java.util.List[Val] = args.asJava

      function.getFunction().apply(argList)
    }

    ValFunction(paramList, f, function.hasVarArgs)
  }

}
